open-craft/dalite-cutover-scripts
# Cutover scripts for dalite.

# Deployment testing

As a part of testing this task you'll need to deploy dalite.ng using scripts from this repository: https://github.com/open-craft/deployment-secrets-services . Since this change is not yet accepted by Xavier, so you'll need to check out `jbzdak/secrets`. 

Secrets repository checks out public repository as a submodule, and there are readmes in both of them. Basically you'll need to deploy new copy of dalite, with following things separate from the one already deployed: 
- Tarsnap keys should be different (this our favourite backup solution)
- Dead Man's Snitches should be different
- Database should be different (but since database provisioning is not part of this task, I created a database for you --- check the ticket for details. 

Probably just make a copy of private vars and edit there. 

Then you should deploy new version of the director and test migration procedure from scratch. 
# Cutover scripts
### How this works

Moving dalite from one instance to other consists of doing following things: 
1. Login from the director to old dalite, and perform mysqldump. 
2. Truncate database of new dalite 
3. Load the data dump to mysql 
4. Download s3 container for media uploads
5. Upload data to swift container 
6. Rsync the logs from `/home/log-downloader` from one machine to another. 
### Test deployment

I don't really think that testing scripts that deploy director instance will be helpful, however to do this you need to do following thing to test creation of new director instance: 
1. Checkout this repository 
2. Download private vars (attached to the OC ticket) 
3. Set IP in hosts (create your own instance, with at least 20GB disk)
4. Download vault password from the usual place where we store secrets and save it as `.vault-pass`
5. `ansible-galaxy install -r requirements.yml -f`
6. `ansible-playbook deploy-all.yml --vault-password-file .vault-pass -u administarative-user --extra-vars @private-extra-vars.yaml`
### Test server state
1. Add following mapping to your hosts file: `149.202.187.85 dalite.harvardx.harvard.edu`
2. Open old dalite at https://52.11.239.205/ also open new dalite. 
3. Try to confirm that old and new states for dalite are the same. 
4. Add some assignments and questions in new dalite, add some static files to those questions, 
5. Try to connect from lms on your devstack to the new dalite instance. 
### Test cutover procedure
1. Read the code in this repository and check the scripts that move data. 
2. Login to director instance (IP of the director instance is here: `149.202.184.160`)
3. Run `run-all.sh`
4. Check if all questions you added in the last step disappeared, and both dalites are now perfect copies. 
# Author concerns
1. Database is copied in a peculiar way, we dump data without schema from production server. Then we truncate all tables on new server, and then we load all data on the new server. This is due to the fact, that due to version mismatch between old and new seystems, we hit 767 index length limit on mysql. . .  This is addressed by backwards-incompatible change [in this pr](https://github.com/open-craft/django-lti-tool-provider/pull/6), and I decided that its better than touching production original dalite production instance days before it is torn down.
